sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java

```diff
@@ -1336,14 +1334,40 @@ public boolean getNeedsMessageId() {
 
   @Override
   public PCollection<PubsubMessage> expand(PBegin input) {
-    return input
-        .getPipeline()
-        .begin()
-        .apply(Read.from(new PubsubSource(this)))
-        .apply(
-            "PubsubUnboundedSource.Stats",
-            ParDo.of(
-                new StatsFn(pubsubFactory, subscription, topic, timestampAttribute, idAttribute)));
+    SerializableFunction<byte[], PubsubMessage> function;
+    if (getNeedsAttributes() || getNeedsMessageId()) {
+      function = new PubsubMessages.ParsePubsubMessageProtoAsPayload();
+    } else {
+      function = new DeserializeBytesIntoPubsubMessagePayloadOnly();
+    }
+    PCollection<PubsubMessage> messages =
+        input
+            .getPipeline()
+            .begin()
+            .apply(Read.from(new PubsubSource(this)))
+            .apply(
+                "MapBytesToPubsubMessages",
+                MapElements.into(TypeDescriptor.of(PubsubMessage.class)).via(function));
+    if (usesStatsFn(input.getPipeline().getOptions())) {
+      checkArgument(false);
```

Comment:
Now that I look more closely in the morning it seems there are debugging items left in.

Reply:
Yeah...I'll remove these and fix the tests. Will let you know it's ready.